File: api/src/services/cdn-client.ts
/**
 * CDN Client — Axios wrapper for uploading images to the CDN server.
 *
 * Used by the Core API to proxy image uploads from the frontend/Flutter
 * to the CDN server (Port 3001) with the CDN_API_KEY.
 */

import axios, { AxiosInstance } from "axios";
import { env } from "../config/env";
import FormData from "form-data";

class CdnClient {
    private client: AxiosInstance;

    constructor() {
        this.client = axios.create({
            baseURL: env.CDN_BASE_URL,
            timeout: 30000, // 30 seconds — image processing can be slow
            headers: {
                "x-cdn-key": env.CDN_API_KEY,
            },
        });
    }

    /**
     * Upload an image buffer to the CDN.
     *
     * @param buffer   - Raw image buffer
     * @param filename - Original filename (for Content-Disposition)
     * @param slug     - SEO-friendly slug for the output filename
     * @returns CDN response with url, thumbUrl, etc.
     */
    async uploadImage(
        buffer: Buffer,
        filename: string,
        slug: string
    ): Promise<{
        url: string;
        thumbUrl: string;
        filename: string;
        width: number;
        height: number;
        sizeBytes: number;
    }> {
        const form = new FormData();
        form.append("image", buffer, { filename });
        form.append("slug", slug);

        const response = await this.client.post("/api/media/upload", form, {
            headers: form.getHeaders(),
            maxContentLength: Infinity,
            maxBodyLength: Infinity,
        });

        if (!response.data?.success) {
            throw new Error(response.data?.error || "CDN upload failed");
        }

        return response.data.data;
    }

    /**
     * Delete an image/media from the CDN.
     *
     * @param fileUrl - The full CDN URL of the file to delete
     * @returns true if deleted, false if not found or unsupported
     */
    async deleteImage(fileUrl: string): Promise<boolean> {
        try {
            // Extract the filename from the URL
            const urlObj = new URL(fileUrl);
            const filename = urlObj.pathname.split("/").pop();
            if (!filename) return false;

            const response = await this.client.delete(`/api/media/${filename}`, {
                timeout: 10000,
            });

            return response.data?.success === true;
        } catch (err: any) {
            // If CDN doesn't support delete (404/405), just log and continue
            if (err.response?.status === 404 || err.response?.status === 405) {
                console.warn(`[CDN] Delete not supported or file not found: ${fileUrl}`);
                return false;
            }
            console.warn(`[CDN] Delete failed for ${fileUrl}:`, err.message);
            return false;
        }
    }

    /**
     * Check CDN health.
     */
    async healthCheck(): Promise<boolean> {
        try {
            const res = await this.client.get("/health", { timeout: 5000 });
            return res.data?.status === "ok";
        } catch {
            return false;
        }
    }
}

export const cdnClient = new CdnClient();

File: fast-api/modules/vr_tour/cdn_client.py
"""
CDN Upload Client — Uploads processed panoramas to the Foodies Pakistan CDN.

Mirrors the Node.js cdn-client.ts behavior:
- POST /api/media/upload with x-cdn-key header
- Uploads panorama + thumbnail as multipart/form-data
- Returns CDN URLs for the uploaded files
"""

import logging
from pathlib import Path

import httpx

from config import settings

logger = logging.getLogger("vr_tour.cdn")


class CdnClient:
    """Async HTTP client for CDN uploads."""

    def __init__(self):
        self.base_url = settings.CDN_BASE_URL.rstrip("/")
        self.api_key = settings.CDN_API_KEY
        self.timeout = 120.0  # 2 min — panoramas can be large

    async def upload_image(
        self,
        file_path: str,
        filename: str,
        slug: str,
    ) -> dict:
        """
        Upload an image file to the CDN.

        Args:
            file_path: Path to the local image file
            filename: Name for the uploaded file
            slug: SEO-friendly slug for the output path

        Returns:
            Dict with 'url', 'thumbUrl', 'filename', 'width', 'height', 'sizeBytes'
        """
        url = f"{self.base_url}/api/media/upload"

        with open(file_path, "rb") as f:
            file_bytes = f.read()

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.post(
                url,
                headers={"x-cdn-key": self.api_key},
                files={"image": (filename, file_bytes, "image/webp")},
                data={"slug": slug},
            )

            if response.status_code != 200:
                logger.error(f"CDN upload failed: {response.status_code} — {response.text}")
                raise Exception(f"CDN upload failed with status {response.status_code}")

            data = response.json()
            if not data.get("success"):
                raise Exception(data.get("error", "CDN upload returned error"))

            logger.info(f"CDN upload success: {data['data']['url']}")
            return data["data"]

    async def health_check(self) -> bool:
        """Check CDN connectivity."""
        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                r = await client.get(
                    f"{self.base_url}/health",
                    headers={"x-cdn-key": self.api_key},
                )
                return r.json().get("status") == "ok"
        except Exception:
            return False


cdn_client = CdnClient()
